### PR TITLE
Fix cases when intersect / except is not checked

### LINKED ISSUE #30803: Segfault on query analysis when using INTERSECT
hi,

    the Clickhouse had a crash exception When I didn't do nothing, Clickhouse has a table engine and dictionary based on Postgres database，The error reporting information is as follows:

### environments:
```
ch: version 21.9.2.17 (official build), build id: E4F05ABB2100332308613C22030A03F5A4621821
```

### err log:
```
 {} <Fatal> BaseDaemon: ########################################
{} <Fatal> BaseDaemon: (version 21.9.2.17 (official build), build id: E4F05ABB2100332308613C22030A03F5A4621821) (from thread 11581) (query_id: 0febacc4-6e53-43b0-bf79-d091407d6535) Received signal Segmentation fault (
11)
 {} <Fatal> BaseDaemon: Address: 0x40 Access: read. Address not mapped to object.
 {} <Fatal> BaseDaemon: Stack trace: 0x11bcf1a9
 {} <Fatal> BaseDaemon: 1. DB::ASTSelectQuery::withFill() const @ 0x11bcf1a9 in /usr/lib/debug/.build-id/e4/f05abb2100332308613c22030a03f5a4621821.debug
 {} <Fatal> BaseDaemon: Checksum of the binary: B5C3F64B513CF1F98A9EACDC3AE921EE, integrity check passed.
 {} <Fatal> Application: Child process was terminated by signal 11.
```

### crash log:
```
# 
# trace full
['1. DB::ASTSelectQuery::withFill() const @ 0x11bcf1a9 in /usr/lib/debug/.build-id/e4/f05abb2100332308613c22030a03f5a4621821.debug']
```

### PR DESCRIPTION
**Pull request by kssenii**
Changelog category (leave one):
- Bug Fix (user-visible misbehaviour in official stable or prestable release)

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix some corner cases with intersect/except. Closes https://github.com/ClickHouse/ClickHouse/issues/30803.

### CODE REVIEW

File: src/Interpreters/ApplyWithGlobalVisitor.cpp [hunk 1]
```diff
@@ -40,14 +41,40 @@ void ApplyWithGlobalVisitor::visit(
...
     }
 }
 
 void ApplyWithGlobalVisitor::visit(ASTPtr & ast)
 {
     if (ASTSelectWithUnionQuery * node_union = ast->as<ASTSelectWithUnionQuery>())
     {
-        if (auto * first_select = node_union->list_of_selects->children[0]->as<ASTSelectQuery>())
+        auto * first_select = typeid_cast<ASTSelectQuery *>(node_union->list_of_selects->children[0].get());
+        if (first_select)
```
Comment (evillique) [inline src/Interpreters/ApplyWithGlobalVisitor.cpp:77]: Why change here is needed?

File: src/Parsers/ASTSelectIntersectExceptQuery.h [hunk 2]
```diff
@@ -1,29 +1,32 @@
 #pragma once
 
 #include <Parsers/ASTQueryWithOutput.h>
```
Comment (evillique) [inline src/Parsers/ASTSelectIntersectExceptQuery.h:3]: Could probably be removed
```suggestion
```

File: src/Interpreters/PredicateRewriteVisitor.cpp [hunk 3]
```diff
@@ -37,13 +38,21 @@ void PredicateRewriteVisitorData::visit(ASTSelectWithUnionQuery & union_select_q
     for (size_t index = 0; index < internal_select_list.size(); ++index)
     {
         if (auto * child_union = internal_select_list[index]->as<ASTSelectWithUnionQuery>())
+        {
             visit(*child_union, internal_select_list[index]);
-        else
+        }
+        else if (auto * child_intersect_except = internal_select_list[index]->as<ASTSelectIntersectExceptQuery>())
```
Comment (evillique) [inline src/Interpreters/PredicateRewriteVisitor.cpp:44]: Unused variable or unfinished branch :(

File: src/Client/QueryFuzzer.cpp [hunk 4]
```diff
@@ -447,6 +448,11 @@ void QueryFuzzer::fuzz(ASTPtr & ast)
     {
         fuzz(with_union->list_of_selects);
     }
+    if (auto * with_intersect_except = typeid_cast<ASTSelectIntersectExceptQuery *>(ast.get()))
```
Comment (evillique) [inline src/Client/QueryFuzzer.cpp:451]:
```suggestion
    }
    else if (auto * with_intersect_except = typeid_cast<ASTSelectIntersectExceptQuery *>(ast.get()))
```

Comment (evillique) [inline src/Client/QueryFuzzer.cpp:451]:
```suggestion
    else if (auto * with_intersect_except = typeid_cast<ASTSelectIntersectExceptQuery *>(ast.get()))
```

**Review by evillique — APPROVED** [review]

**Comment by kssenii** [comment]
`Functional stateless tests (release, DatabaseReplicated) — Timeout` -- intersecting parts, told Sasha about it.
`Functional stateless tests (release, wide parts enabled) — Timeout` -- segfault in MergePlainMergeTreeTask, told Nikita about it.
failed `Integration tests`:
1. test_storage_hdfs/test.py::test_read_files_with_spaces -- fixed in https://github.com/ClickHouse/ClickHouse/pull/31250.
2. test_storage_kafka -- something strange.

**Pull request merged**